[WIP] Adding docs form symfony form creation from module (#1060)
PrestaShop/docs: src/content/1.7/modules/symfonyform/_index.md

---
title: Symfony form in module
weight: 10
---

# Symfony forms

> One of key objectives in PrestaShop's migration to Symfony is migrating legacy PrestaShop forms to architecture based on Symfony Forms. https://symfony.com/doc/3.4/forms.html 

Reviewer: Shouldn't the links in documention refer to Symfony version 4.4 ?

Module developers can use Symfony Forms in they modules as well. Symfony documentation gives a lot of useful information about forms, but there are some things that are specific to PrestaShop.

# Getting started

In order to use Symfony form in your module you require:

1) A form type
2) A template to display that form type

> 3) A way to handle that form.

Reviewer: ```suggestion
3) A form handler responsible for providing the form data and processing its submission
```

It still uses the same FormFactory as Symfony does, but it also sets data to the form and calls the hook specific to that form. 

You cam find detailed documentation to how Settings form works here:
https://devdocs.prestashop.com/1.7/development/architecture/migration-guide/forms/settings-forms/

And in this example module you can find an implementation:
https://github.com/PrestaShop/example-modules/tree/master/demosymfonyform

##### CRUD Form

CRUD form is more complicated because it works with logic objects. 
You can use ``PrestaShop\PrestaShop\Core\Form\IdentifiableObject\Builder\FormBuilder`` to get the form. 
Form builder will require form type and data provider. Data provider is needed in order to fill form with data if it's an edit form. 
Data provider must implement FormDataProviderInterface where you have functions ``getData($id)`` (retrieve data based on id) and ``getDefaultData()``.

CRUD form handler:
https://devdocs.prestashop.com/1.7/development/architecture/migration-guide/forms/crud-forms/

After setting up form handler. You can use it to retrieve the form.

```php
$textForm = $textFormDataHandler->getForm();

> ```

Reviewer: How is this form handler defined in module configuration ?
How do we get its instance ? by dependency injection ?

> After retrieving form you can use it to get create form view which then can be used in twig template to render said form.

Reviewer: ```suggestion
After retrieving form you can use it to get the form view which then can be used in twig template to render said form.
```

```php
$this->render('@Modules/demosymfonyform/views/templates/admin/form.html.twig', [
            'demoConfigurationTextForm' => $textForm->createView(),
        ]);
````

In twig before PrestaShop 1.7.8 you needed to define every field for best results. 
This allows for greater control over the form as you can define HTML exactly as you need it, but it's requires extra work if all you need is a standard form.

```twig
{% extends '@PrestaShop/Admin/layout.html.twig' %}

> {# PrestaShop made some modifications to the way forms are displayed to work well with PrestaShop in order to benefit from those you need to use ui kit as theme#}

Reviewer: ```suggestion
{# PrestaShop made a few modifications to original Symfony templates in order to properly display forms in PrestaShop context, and to benefit from UI kits such as theme#}
```

> {% form_theme demoConfigurationTextForm 'PrestaShopBundle:Admin/TwigTemplateForm:prestashop_ui_kit.html.twig' %}

Reviewer: Why form theming isn't applied by default in Symfony configuration ?
https://symfony.com/doc/4.4/form/form_themes.html#applying-themes-to-all-forms
```
# config/packages/twig.yaml
twig:
    form_themes: ['PrestaShopBundle:Admin/TwigTemplateForm:prestashop_ui_kit.html.twig']
```

    <h3 class="card-header">
      <i class="material-icons">settings</i> {{ 'Text form types'|trans({}, 'Modules.DemoSymfonyForm.Admin') }}
    </h3>
    <div class="card-block row">
      <div class="card-text">
        {{ form_widget(demoConfigurationTextForm) }}
      </div>
    </div>
    <div class="card-footer">
      <div class="d-flex justify-content-end">
        <button class="btn btn-primary float-right" id="save-button">
          {{ 'Save'|trans({}, 'Admin.Actions') }}
        </button>
      </div>
    </div>
  </div>
  {{ form_end(demoConfigurationTextForm) }}
{% endblock %}
```

{{% notice note %}}
Some form types may require JS components. You can always check documentation for specific form type if it requires one.
You can find out more about JS components here: https://devdocs.prestashop.com/1.7/development/components/global-components/
{{% /notice %}}

## Form handling

At this point you should have form rendered successfully, so it's time to handle the form.

Usually you handle the form in the same action that renders form tempalate. You just need to check if the form is submitted using ``isSubmitted`` function.
```php
if ($textForm->isSubmitted()) {
}

> ``` 

Reviewer: Before this step, should the documentation describe how to handle the form from the current request ?
```php
$textForm->handleRequest($request);
```

You should also check if the form is valid if you are using constraints.

```php
if ($textForm->isSubmitted() && $textForm->isValid()) {
        
}
```

> Inside this ``if`` you can handle the form. You can data submitted in the form using forms ``getData`` and you can handle that data any way you want.

Reviewer: ```suggestion
Inside this ``if`` structure you can handle the form submission. You can access the data submitted in the form using its method ``getData()``, and then handle the data in any way that fits your need.
```

It's advised to use PrestaShops form handlers as it saves you some work and also calls hooks so other modules can interact with your forms if needed. There are multiple handlers, so it depends on what type of form do you have.

##### Settings form

Settings form in PrestaShop can be handled with same handler it was created: ``PrestaShop\PrestaShop\Core\Form\Handler.php``. You need to use it's ``save`` function.
 ```php
 if ($textForm->isSubmitted() && $textForm->isValid()) {
            /** You can return array of errors in form handler and they can be displayed to user with flashErrors */
            $errors = $textFormDataHandler->save($textForm->getData());

            if (empty($errors)) {
                $this->addFlash('success', $this->trans('Successful update.', 'Admin.Notifications.Success'));

                return $this->redirectToRoute('demo_configuration_form');
            }

            $this->flashErrors($errors);
        }
```

> When defining service for Handler one of arguments is a class implements interface ``PrestaShop\PrestaShop\Core\Form\FormDataProviderInterface``, 

Reviewer: ```suggestion
When defining the handler in configuration, one of its arguments has to be the class that implements the interface ``PrestaShop\PrestaShop\Core\Form\FormDataProviderInterface``, 
```

> When defining service for Handler one of arguments is a class implements interface ``PrestaShop\PrestaShop\Core\Form\FormDataProviderInterface``, 

Reviewer: Configuration example ?

> In that class you can use ``setData`` function to save data and pass back any errors, so it can also be used for validation.

Reviewer: ```suggestion
In this class you can use ``setData()`` method to store data and return any errors, so it can also be used for validation.
```

```php
class DemoConfigurationChoiceFormDataProvider implements FormDataProviderInterface
{

>     /** If you want to prefill your from with data when it's loaded(previous values if it's setting form) you can return that data here */

Reviewer: ```suggestion
    /** If you want to prefill data in your form as soon as it gets loaded (previous values if it's setting form), then you can return that data here */
```

    public function getData(): array
    {

>         return $this->getData();

Reviewer: Looks like a recursion to infinite loop here

    }

    public function setData(array $data): array
    {
        $this->saveData($data);
        /** You can return array of errors if something gfoes wrong */
        return ['Something went wrong'];

    }
}
```

##### CRUD Form

> To handle CRUD form PrestaShop uses ``PrestaShop\PrestaShop\Core\Form\IdentifiableObject\Handler``

Reviewer: ```suggestion
To handle CRUD form you should use the dedicated handler provided by PrestaShop: ``PrestaShop\PrestaShop\Core\Form\IdentifiableObject\Handler``
```

> You can use ``FormHandlerFactory`` ``create`` function in order to create the handler. Primary argument you need here is DataHandler which would be responsible for saving your data.

Reviewer: ```suggestion
You can use ``FormHandlerFactory::create()`` method in order to create the handler from its definition in configuration. The data handler is reqponsible for saving your data, and it has to be the first argument to define in said configuration.
```

> It needs to implement ``FormDataHandlerInterface`` and can be passed to your form handler in a service.

Reviewer: ```suggestion
Such data handler has to implement the interface ``FormDataHandlerInterface``, and it is injected to your form handler as a service.
```

```yaml
  prestashop.core.form.identifiable_object.handler.feature_form_handler:
    class: 'PrestaShop\PrestaShop\Core\Form\IdentifiableObject\Handler\FormHandler'
    factory: 'prestashop.core.form.identifiable_object.handler.form_handler_factory:create'
    arguments:
      #  Your data handler
      - '@prestashop.core.form.identifiable_object.data_handler.feature_form_data_handler'
```

> Data handler has to functions ``create`` and ``update`` where you need to define logic for your object. 

Reviewer: ```suggestion
Data handler has to implement methods ``create`` and ``update`` where you need to define the "logic" of your object. 
```

In controller use either ``FormHandler->handle`` or ``FormHandler->handleFor``.

> ``handleFor`` requires you to have pass an ID as it's used to update object, while ``handle`` is used to create one.

Reviewer: ```suggestion
``handleFor`` requires you to provide the object ID in order to update said object, while ``handle`` is used to create a new object.
```

##### Using different form action

> In some cases you want to use new action for form saving (for example if you have multiple forms in one page and you want each to have separate action)

Reviewer: ```suggestion
In some cases you might need to implement a dedicated action to process form submission (for example if you have multiple forms in one page to get handled by distinct actions)
```

> In that case you can define route in template file, you need to modify a ``form_start`` and add path to your action.

Reviewer: ```suggestion
In such case you can override the routes to submit to in the template file, and provide the path to the related action as an option in the ``form_start`` function.
```

```twig
 {{ form_start(demoConfigurationChoiceForm, {attr : {class: 'form'}, action: path('demo_configuration_choices_form_save') }) }}
```

> In that case when you are done handling the form you will need to redirect back to original form. You can pass success or error messages using controller function
> ``addFlash``. You can't use constraints when handling form like this since after redirecting you won't see any error messages. ``$textForm->isValid()`` will still return negative however so it's advise to not to use it, validate during form handling and add errors using ``addFlash``.  

Reviewer: ```suggestion
Once you are done handling the form you will have to redirect the navigation back to the initial page. You can notify success or error messages using controller function ``addFlash``. You can't use constraints when handling a form this way because redirection will hide any error messages. ``$textForm->isValid()`` will still return negative however so it's advisable to not to use it. Instead validate during form handling and add errors using ``addFlash``.  
```

> ``addFlash``. You can't use constraints when handling form like this since after redirecting you won't see any error messages. ``$textForm->isValid()`` will still return negative however so it's advise to not to use it, validate during form handling and add errors using ``addFlash``.  

Reviewer: I strongly disagree here. We should advise to redirect only when the form is valid and properly handled. Otherwise we should render the page in the current action/request in order to display current form errors.
This is the basic behaviour promoted by Symfony documentations.

##### 

```php
<?php
header('Expires: Mon, 26 Jul 1997 05:00:00 GMT');
header('Last-Modified: ' . gmdate('D, d M Y H:i:s') . ' GMT');
header('Cache-Control: no-store, no-cache, must-revalidate');
header('Cache-Control: post-check=0, pre-check=0', false);
header('Pragma: no-cache');
header('Location: ../');
exit;

> ```

Reviewer: This is the default behavior of Symfony redirection. No need to handle it manually.
